### PR TITLE
Add hotReloadInitialDevFSSyncMilliseconds to track how long user have…

### PR DESCRIPTION
**Pull request by aam**
… to wait before being able to do first hot reload.

This stat is significantly different between existing and preview-dart-2 setting at the moment(for the latter, this stat is ~3x slower: 13s vs ~4s). The reason for difference is that first time we do devfs sync we also initialize incremental Dart Frontend - essentially parse the whole application at that point.

### CODE REVIEW

**Comment by a-siva** [comment]
The plan is to do the initial compile using the incremental Dart Frontend at which point we won't have to parse the whole application again during devfs sync. Considering this do you still see a point in adding this stat?

**Comment by aam** [comment]
I think so, yes. If we are able to skip compilation at this step this stat will show an improvement.

Completely getting rid of compilation at this step might not be quite straightforward.
Basically we need to consider the case when flx does need to be rebuilt before hot run - there we do want to somehow keep incremental compiler used to build flx in memory and keep it for hot run too.
But there is also a case where flx does NOT have to be rebuilt before hot run - there we don't have a choice but to initialize incremental compiler when user runs the app.

IIUC potentially incremental compiler can skip parsing altogether if we start storing it's state on the file system rather in memory(which we currently use). @scheglov @sigmundch hopefully can confirm that.

**Comment by Hixie** [comment]
Having benchmarks is always good. We catch a ton of unintentional regressions with the ones we have.

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 1]
```diff
@@ -128,7 +128,17 @@ class HotRunner extends ResidentRunner {
...
+      initialUpdateDevFSsTimer = new Stopwatch();
+      initialUpdateDevFSsTimer.start();
+    }
     final bool devfsResult = await _updateDevFS();
+    if (benchmarkMode) {
+      initialUpdateDevFSsTimer.stop();
+      benchmarkData['hotReloadInitialDevFSSyncMilliseconds'] =
+          initialUpdateDevFSsTimer.elapsed.inMilliseconds;
+    }
     if (!devfsResult) {
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/run_hot.dart]: nit: while you're here, remove the braces on this one line if statement :-)

**Comment by Hixie** [comment]
LGTM

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 2]
```diff
@@ -128,7 +128,17 @@ class HotRunner extends ResidentRunner {
       printError('Error initializing DevFS: $error');
       return 3;
     }
+    Stopwatch initialUpdateDevFSsTimer;
+    if (benchmarkMode) {
+      initialUpdateDevFSsTimer = new Stopwatch();
+      initialUpdateDevFSsTimer.start();
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/run_hot.dart]: nit: it is common to use `..start()` to start a timer on the same line as where it is initialized. In fact, I'd even remove the `if (benchmarkMode)` and always started a `Stopwatch`. It makes no difference performance-wise, and turns 5 line from 131 to 135 into one line:

```dart
final Stopwatch initialiUpdateDevFSsTimer = new Stopwatch()..start();
```

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 3]
```diff
@@ -128,7 +128,17 @@ class HotRunner extends ResidentRunner {
...
+    }
     final bool devfsResult = await _updateDevFS();
+    if (benchmarkMode) {
+      initialUpdateDevFSsTimer.stop();
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/run_hot.dart]: nit: you don't need to stop the `Stopwatch` if you do not plan to resume it later.

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 4]
```diff
@@ -128,7 +128,17 @@ class HotRunner extends ResidentRunner {
       printError('Error initializing DevFS: $error');
       return 3;
     }
+    Stopwatch initialUpdateDevFSsTimer;
+    if (benchmarkMode) {
+      initialUpdateDevFSsTimer = new Stopwatch();
+      initialUpdateDevFSsTimer.start();
+    }
     final bool devfsResult = await _updateDevFS();
+    if (benchmarkMode) {
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/run_hot.dart]: I know you did not start this trend, but I'm not sure why we're recording benchmark data conditionally. It costs nothing. We should probably only consider `benchmarkMode` when outputting this data. I think removing some of these conditionals could simplify our code.

**Review by yjbanov — APPROVED** [review]
Only nitpicks. LGTM.

**Comment by aam** [comment]
Thanks @Hixie and @yjbanov ! I incorporated the changes you suggested. PTAL!

**Review by yjbanov — APPROVED** [review]
LGTM.

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 5]
```diff
@@ -362,10 +363,8 @@ class HotRunner extends ResidentRunner {
         '${getElapsedAsMilliseconds(restartTimer.elapsed)}.');
     // We are now running from sources.
     _runningFromSnapshot = false;
-    if (benchmarkMode) {
-      benchmarkData['hotRestartMillisecondsToFrame'] =
-          restartTimer.elapsed.inMilliseconds;
-    }
+    benchmarkData['hotRestartMillisecondsToFrame'] =
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/run_hot.dart]: Thanks for cleaning up all the other places!

**Comment by sigmundch** [comment]
> IIUC potentially incremental compiler can skip parsing altogether if we start storing it's state on the file system rather in memory(which we currently use). @scheglov @sigmundch hopefully can confirm that.

correct - if you use a persistent byte store, we should be able to reuse data from previous compilations when computing the initial .dill file.

**Pull request merged**